### ISSUE TITLE
Generate library benchmarks dynamically.

### ISSUE THREAD

**Issue by infogulch** [iai-callgrind/iai-callgrind#34]
Thanks to your suggestions I made some improvements to the benchmark code for scryer-prolog.

* [setup.rs](https://github.com/mthom/scryer-prolog/blob/master/benches/setup.rs)
* [run_iai.rs](https://github.com/mthom/scryer-prolog/blob/master/benches/run_iai.rs)

But I already ran into an issue where I forgot to create a `#[library_benchmark]` for a corresponding benchmark definition in my own format in `setup.rs`. [original addition](https://github.com/mthom/scryer-prolog/commit/98d8b00d7a337dad42a7cdca3e717999e857affa) / [add it to iai benches list later](https://github.com/mthom/scryer-prolog/pull/2169/commits/7e313d169efb6b20ef3ab4ea046a228483d3edec)

This isn't an issue for [criterion benchmarks](https://github.com/mthom/scryer-prolog/blob/master/benches/run_criterion.rs) because there I iterate through the list dynamically.

Any thoughts on how I can run the iai benchmarks directly from my custom definition list?

**Comment by gamma0987** [comment]
> Thanks to your suggestions I made some improvements to the benchmark code for scryer-prolog.

Sure, no problem :)

The dynamics of criterion.rs you have with bench_iter aren't possible for library benchmarks in iai-callgrind. Especially, the builder like api for library benchmarks like that from criterion is not transferable to iai-callgrind. That's also the reason why there are different schemes for binary benchmarks and library benchmarks. I'm open for suggestions but, I'm going to have another look into it when I have more time for it.

Having a quick look at your iai benchmarks, imho the best you can currently do is listing all the setups manually. If you like to, you can merge the similar setups in a single `#[library_benchmark]`, so adding a new `#[bench::...]` is almost just copy and paste and it's harder to overlook a missing setup:

```rust
use iai_callgrind::{library_benchmark, library_benchmark_group, main};

mod setup;

#[library_benchmark]
#[bench::count_edges_short(setup::prolog_benches()["count_edges_short"].setup())]
#[bench::numlist_short(setup::prolog_benches()["numlist_short"].setup())]
# ...
fn bench(mut run: impl FnMut()) {
    run();
}

library_benchmark_group!(
    name = bench_group;
    benchmarks = bench
);
main!(library_benchmark_groups = bench_group);
```

Just a note, but I saw that in your `run` closure you asserted the result at the end. This skews your benchmarks, especially the ones for iai-callgrind. I can understand the temptation to test and benchmark in one go, but if possible, I would separate concerns here and use separate tests where you confirm the correctness of the results. Then, in your benchmarks, I would stick to an `unwrap` and return the unwrapped result of the benchmarking function something like this

```rust
# ...
fn bench(mut run: impl FnMut()) -> Whatever {
    run().unwrap()
}
```

**Comment by infogulch** [comment]
> you can merge the similar setups in a single `#[library_benchmark]`

Good idea, that helps a bit.

I tried to minimize its impact of the assert, but you're right there's no reason not to move that to a test.

Thanks for your input!

**Comment by infogulch** [comment]
I know I closed this, but I'm having a hell of a time trying to make it actually run the test fn. I tried putting them in setup.rs, under a `#[cfg(test)] mod test { ...` but I can't find a command that will actually run them.

This is the commit where I added them: https://github.com/mthom/scryer-prolog/pull/2169/commits/76a133c8a19249588b3125f02b011445c8eeb199

I've tried:

* `cargo test --all` -- doesn't run
* `cargo test --tests` -- doesn't find the test
* `cargo test validate_benchmarks` -- (the name of the test function) doesn't find the test
* `cargo test --tests --all-targets` -- actually runs the benchmark binaries which is not helpful
* `cargo test --all-targets validate_benchmarks` -- actually runs the benchmark binaries which is not helpful
* `cargo bench --all` -- this fails for some reason
* `cargo bench --bench run_iai` and `cargo bench --bench run_iai` neither actually run the tests

I'm at a bit of a loss.
